peg-macros/translate.rs

```diff
@@ -158,11 +158,33 @@ fn make_parse_state(grammar: &Grammar) -> TokenStream {
     let mut cache_fields_def: Vec<TokenStream> = Vec::new();
     let mut cache_fields: Vec<Ident> = Vec::new();
     for rule in grammar.iter_rules() {
-        if rule.cache.is_some() && rule.params.is_empty() && rule.ty_params.is_none() {
+        if rule.cache.is_some() && rule.ty_params.is_none() {
             let name = format_ident!("{}_cache", rule.name);
             let ret_ty = rule.ret_type.clone().unwrap_or_else(|| quote!(()));
+
+            let key = if rule.params.is_empty() {
+                quote_spanned!(span=> usize)
+            } else {
+                let param_ty = rule.params.iter().map(|param| match &param.ty {
+                    RuleParamTy::Rust(ty) => ty
+                        .clone()
+                        .into_iter()
+                        .filter(|tt| match tt {
+                            // TODO: This needs to do more; at the least, there
+                            // may be lifetimes. Doing this correctly for every
+                            // edge case probably requires syn, which is a heavy
+                            // dependency.
+                            TokenTree::Punct(punct) => punct.as_char() != '&',
+                            _ => true,
+                        })
+                        .collect(),
+                    RuleParamTy::Rule(ty) => ty.clone(),
```

Comment:
I don't think this will work? This is for a parameter declared as `rule<T>`, which at the Rust level is a closure returning `RuleResult<T>`. To work properly, the cache key would need to be the identity of the closure definition site plus any closed-over upvars, which would be quite difficult to do. So this should probably remain forbidden.

Reply:
You’re right, I improperly assumed that prohibiting generic types was sufficient for this case because I almost never used `rule` with a concrete type. Now that you point it out, it is embarrassingly obvious that `RuleParamTy` appearing here actually means I needed to think about it more :-)